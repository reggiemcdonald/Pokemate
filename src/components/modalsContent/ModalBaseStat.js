import React from "react";
import {
    View,
    Text,
    ScrollView,
    FlatList,
    TouchableOpacity,
    StyleSheet
} from "react-native";
import {StatNameFormats} from "../../library/StringResources";
import InvalidValue from "../../library/errors/InvalidValue";

/**
 * **************************
 * Base Stat Modal Component
 * *************************
 * Displays base stat information
 * Consumes a prop of PokeDataManager and a callback to subscribe to changes in the instantiating parent
 */
export default class ModalBaseStat extends React.Component {

    /**
     * Render the component
     **/
    render() {
        if (this.props.data === undefined) {
            return(
                <View style={modalStyleSheet}>
                    <TouchableOpacity onPress={this.props.closeFunction}>
                        <Text style={{textAlign: "center"}}>Close</Text>
                    </TouchableOpacity>
                </View>
            );
        } else {
            let positivelyAffectingMovesTitle = "Moves: Positive";
            const negativelyAffectingMovesTitle = "Moves: Negative";
            const positivelyAffectingNaturesTitle = "Natures: Positive";
            const negativelyAffectingNaturesTitle = "Natures: Negative";
            return (
                <View>
                    <View style={modalStyleSheet.modalTitleBar}>
                        <Text testID={"testIdName"}
                              style={[modalStyleSheet.modalTitleText, modalStyleSheet.textGreedy]}>
                            {StatNameFormats[this.props.data.name]}
                        </Text>
                        <TouchableOpacity onPress={this.props.closeFunction}>
                            <Text style={modalStyleSheet.closeButton}>Close</Text>
                        </TouchableOpacity>
                    </View>
                    <ScrollView style={modalStyleSheet.modalViewContainer}>
                        <View style={{flexDirection: "row", alignItems: "flex-start"}}>
                            {this._renderFlatList(this.props.data.affectingMoves.positive,
                                "testIdMovesPositive,", positivelyAffectingMovesTitle)}
                            {this._renderFlatList(this.props.data.affectingMoves.negative,
                            "testIdMovesNegative", negativelyAffectingMovesTitle)}
                        </View>
                        <View style={{flexDirection: "row", alignItems: "flex-start"}}>
                            {this._renderFlatList(this.props.data.affectingNatures.positive,
                            "testIdNaturesPositive", positivelyAffectingNaturesTitle)}
                            {this._renderFlatList(this.props.data.affectingNatures.negative,
                            "testIdNaturesNegative", negativelyAffectingNaturesTitle)}
                        </View>
                        {/*TODO: Must add affecting characteristics*/}
                    </ScrollView>
                </View>
            );
        }
    }

    /**
     * Returns true when the component data is intact, and
     * throws InvalidValue when it is not
     * @private
     */
    _validateData()  {
        try {
            this._validateBaseStat();
            this._validateAffectingElements(this.props.data.affectingMoves);
            this._validateAffectingElements(this.props.data.affectingNatures);
            this._validateAffectingCharacteristics();
        } catch (err) {
            throw err;
        }
    }

    /**
     * Returns true if the base stat is valid
     * @private
     */
    _validateBaseStat() {
        if (this.props.data.name === undefined || !StatNameFormats.hasOwnProperty(this.props.data.name)) {
            throw new InvalidValue(this.props.data.name);
        } else {
            return true;
        }
    }

    /**
     * Returns true if the given elements is {
     *     postiive: object[],
     *     negative: object[]
     * }
     * with each object in the object[] consisting of {
     *     name: {string},
     *     change?: {number}
     * }
     * Otherwise throws InvalidValue
     * @param elements
     * @returns {boolean}
     * @private
     */
    _validateAffectingElements(elements) {
        if (elements === undefined) {
            throw new InvalidValue(elements);
        }
        for (let element of elements.positive) {
            if (!element.hasOwnProperty("name") || typeof element.name !== "string") {
                // TODO: Update error messages
                throw new InvalidValue("Bad value given in affecting elements of a base stat");
            }
            if (element.hasOwnProperty("change") && typeof element.change !== "number") {
                throw new InvalidValue("Bad value given in affecting elements of a base stat");
            }
        }
        for (let element of elements.negative) {
            if (!element.hasOwnProperty("name") || typeof element.name !== "string") {
                // TODO: Update error messages
                throw new InvalidValue("Bad value given in affecting elements of a base stat");
            }
            if (element.hasOwnProperty("change") && typeof element.change !== "number") {
                throw new InvalidValue("Bad value given in affecting elements of a base stat");
            }
        }
        return true;
    }

    /**
     * Returns true if the affecting characteristics array is defined in the props
     * and is of type {string[]}
     * @private
     */
    _validateAffectingCharacteristics() {
        if (this.props.data.characteristics === undefined) {
            throw new InvalidValue(this.props.data.characteristics);
        }
        for (let characteristic of this.props.data.characteristics) {
            if (typeof characteristic !== "string") {
                // TODO: Update error messages
                throw new InvalidValue("value given as a base stat affecting characteristic");
            }
        }
    }

    /**
     * Renders a flat list with the given data and title
     * @param data: [{name: string, change?: string}], the data to be inserted into the flat list
     * @param testingId: string, the ID for testing purposes
     * @param title: string, The title to be given to this flat list
     * @private
     */
    _renderFlatList(data, testingId, title) {
        return (
            <View style={modalStyleSheet.flatListContainer}>
                <Text style={modalStyleSheet.flatListTitleText}>
                    {title}
                </Text>
                <FlatList
                    testId={testingId}
                    data={data}
                    renderItem={({item,index}) => this._renderFlatListItem(item,index)}
                    keyExtractor={(item,index)=>(item.name)}
                    style={modalStyleSheet.modalBaseStatFlatList}
                />
            </View>
        )
    }

    /**
     * Renders an item for the flat lists of this component
     * @param item
     * @returns {*}
     * @private
     */
    _renderFlatListItem(item, index) {
        if (item.hasOwnProperty("change")) {
            return (
                <View style={[modalStyleSheet.flatListTextContainer, {backgroundColor: flatListColors[index % 2]}]}>
                    <Text style={[modalStyleSheet.textGreedy, modalStyleSheet.flatListText]}>{item.name}</Text>
                    <Text style={modalStyleSheet.flatListText}>{item.change}</Text>
                </View>
            );
        } else {
            return (
                <View style={[
                    modalStyleSheet.flatListTextContainer,
                    {backgroundColor: flatListColors[index % 2]}
                ]}>
                    <Text style={modalStyleSheet.flatListText}>{item.name}</Text>
                </View>
            );
        }

    }
}

const modalStyleSheet = StyleSheet.create({
    modalViewContainer: {
        padding: 5,
    },
    modalTitleBar: {
        paddingLeft: 5,
        paddingRight: 5,
        marginTop: 30,
        flexDirection: "row",
        alignItems: "flex-start",
        borderStyle: "solid",
        borderBottomColor: "#929292",
        borderBottomWidth: 1,
    },
    modalTitleText: {
        fontSize: 23,
        fontWeight: "bold",
    },
    flatListContainer: {
        flex:1,
        marginLeft: 5,
        marginRight: 5,
    },
    flatListTitleText: {
        fontSize: 18,
        marginTop: 10,
        marginBottom: 5
    },
    modalBaseStatFlatList: {
        maxHeight: 300,
        borderColor: "black",
        borderWidth: 2,
        borderRadius: 5,
    },
    flatListTextContainer: {
        flexDirection: "row",
        alignItems: "flex-start",
    },
    flatListText: {
        fontSize: 16,
        margin: 5
    },
    textGreedy: {
        flex: 1,
    },
    closeButton: {
        fontSize: 18,
        color: "#007aff"
    }
});

const flatListColors = [
    "transparent",
    "#e6f2ff",
];