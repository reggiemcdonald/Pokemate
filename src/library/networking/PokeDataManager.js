/**
 * ***********************
 * A data management class
 * ***********************
 */
import PokeDataProcessor from "./PokeDataProcessor";
import PromiseInterrupt from "../errors/PromiseInterrupt";

export default class PokeDataManager {
    constructor(existingData?) {
        // TODO
        this.promiseInterrupted = false;
        this.processor = new PokeDataProcessor();
        this.pokeData = {};
        this.orderedEvolutionTree = {};
        this.statData = {};
        this._restoreData(existingData);
    }

    /**
     * Restore data if there is any to restore
     * Otherwise, do nothing
     * @param existingData
     * @private
     */
    _restoreData(existingData) {
        if (existingData) {
            this.pokeData = existingData.pokeData ? existingData.pokeData : {};
            this.orderedEvolutionTree = existingData.orderedEvolutionTree ? existingData.orderedEvolutionTree : {};
            this.statData = existingData.statData ? existingData.statData : {};
        }
    }

    /**
     * Creates a JSON serializable object to handoff data between instances of PokeDataManager
     * return {
     *     pokeData: this.pokeData
     *     orderedEvolutionTree: this.orderedEvolutionTree,
     *     statData: this.statData
     * }
     */
    createHandoff() {
        // TODO: implement the handoff
        return {
            pokeData: this.pokeData,
            orderedEvolutionTree: this.orderedEvolutionTree,
            statData: this.statData
        }
    }

    /**
     * Return the pokemon data if its already been generated
     * Otherwise, generate it, save it, and return it
     * @param name
     * @returns {Promise<any>}
     */
     async getPokemonDetails(name) {
        try {
            this.checkForCancellation();
            if (this.pokeData.hasOwnProperty(name)) {
                return this.pokeData[name];
            } else {
                this.checkForCancellation();
                let pokemon = await this.processor.formDefaultSpeciesData(name);
                this.pokeData[name] = pokemon;
                this.checkForCancellation();
                return pokemon;
            }
        } catch (err) {
            throw err;
        }
    }

    /**
     * Returns the list of pokemon species
     * @returns {Promise<any> | Promise<*>}
     */
    async getListOfPokemon() {
         try {
             this.checkForCancellation();
             let pokemonList = await this.processor.getListOfPokemon();
             return pokemonList;
         } catch (err) {
             return err;
         }
    }

    async getSpriteUrl(name) {
        try {
            this.checkForCancellation();
            if (this.pokeData.hasOwnProperty(name)) {
                return this.pokeData[name].sprite;
            }
            this.checkForCancellation();
            let url = await this.processor.getSpriteUrl(name);
            return url;
        } catch (err) {
            throw err;
        }
    }

    /**
     * Generate an array of evolution objects
     * [{
     *     trigger: string,
     *     triggerConditional: string,
     *     spriteArray: {}
     * }]
     * @param data
     */
    async buildEvolutionChain(evolutionData) {
        try {
            this.checkForCancellation();
            if (this.orderedEvolutionTree[evolutionData.species.name]) {
                return this.orderedEvolutionTree[evolutionData.species.name];
            } else {
                this.checkForCancellation();
                let evolutionChain = await this._buildEvolutionChainHelp(evolutionData, 0);
                this.checkForCancellation();
                this.orderedEvolutionTree[evolutionData.species.name] = evolutionChain;
                return evolutionChain;
            }
        } catch (err) {
            throw err;
        }
    }

    async _buildEvolutionChainHelp(evolution, order) {
        try {
            let evolutionChain = [];
            this.checkForCancellation();
            let evolutionObject = await this._buildSpriteObject(evolution.species.name, order, evolution.evolution_details);
            evolutionChain.push(evolutionObject);
            if (evolution.evolves_to.length === 0) {
                return evolutionChain;
            } else {
                for (let subEvolution of evolution.evolves_to) {
                    this.checkForCancellation();
                    let subEvolutionArr = await this._buildEvolutionChainHelp(subEvolution, order+1);
                    evolutionChain.push(... subEvolutionArr);
                }
                return evolutionChain;
            }
        } catch (err) {
            throw err;
        }
    }



    _getEvolutionTriggerConditional(evolutionDetails) {
        let keys = Object.keys(evolutionDetails);
        let triggerConditionalArr = [];
        for (let key of keys) {
            if (key === "trigger") {
                return triggerConditionalArr;
            } else if (evolutionDetails[key] !== null && evolutionDetails[key] !== false && evolutionDetails[key] !== "") {
                triggerConditionalArr.push({
                    conditional: key,
                    requirement: (typeof evolutionDetails[key] !== "object") ? evolutionDetails[key] : evolutionDetails[key].name
                });
            }
        }
        return null;
    }

    async _buildSpriteObject(species, order, evolutionDetails) {
        try {
            this.checkForCancellation();
            let sprite = await this.getSpriteUrl(species);
            if (evolutionDetails.length === 0) {
                return ({
                    name: species,
                    trigger: null,
                    triggerConditional: null,
                    sprite: sprite,
                    order: order
                });
            } else {
                let trigger = evolutionDetails[0].trigger.name;
                let triggerConditional = this._getEvolutionTriggerConditional(evolutionDetails[0]);
                return ({
                    name: species,
                    trigger: trigger,
                    triggerConditional: triggerConditional,
                    sprite: sprite,
                    order: order
                });
            }
        } catch (err) {
            throw err;
        }
    }

    /**
     * Returns the information on the base stat statName
     * @param statName
     * @returns object
     * {
     *     affectingMoves: {moves},
     *     affectingNatures: {natures},
     *     characteristics: [characteristics],
     *     isBattleOnly: boolean
     * }
     */
    async getBaseStat(statName) {
        try {
            if (this.statData.hasOwnProperty(statName)) {
                this.checkForCancellation();
                return this.statData[statName];
            } else {
                this.checkForCancellation()
                let newStatData = await this.processor.getBaseStatData(statName);
                this.statData[statName] = newStatData;
                this.checkForCancellation();
                return newStatData;
            }
        } catch (err) {
            throw err;
        }
    }

    /**
     * returns true if no cancellation has been requested. Otherwise, triggers a cancellation with a PromiseInterrupt
     * @returns {boolean}
     */
    checkForCancellation() {
        if (this.promiseInterrupted) {
            this.promiseInterrupted = false;
            throw new PromiseInterrupt();
        }
        return true;
    }

    /**
     * Cancels the next promise execution
     */
    cancelPromise() {
        this.processor.cancelPromise();
        this.promiseInterrupted = true;
    }
}