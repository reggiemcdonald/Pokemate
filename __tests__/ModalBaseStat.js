import React from "react";
import renderer from "react-test-renderer";
import Enzyme, { shallow, mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import ModalBaseStat from "../src/components/modalsContent/ModalBaseStat";
import ExpectedResults from "../queryResults/ExpectedResults";
import PokeDataManager from "../src/library/networking/PokeDataManager";
import {Text} from "react-native";

describe("ModalBaseStat: render", () => {
    const manager = new PokeDataManager();
    it("Should render without error when given a valid base stats data structure", () => {
        const tree =
            renderer.create(<ModalBaseStat data={ExpectedResults.baseStatsHp}/>).toJSON();
        expect(tree).toMatchSnapshot();
    });
    it("Should render a novel base stats data structure without error", async () => {
        const data = await manager.getBaseStat("defense");
        const tree =
            renderer.create(<ModalBaseStat data={data}/>).toJSON();
        expect(tree).toMatchSnapshot();
    });
    it("The name should be displayed correctly", () => {
        Enzyme.configure({adapter: new Adapter()});
        const wrapper = shallow(
                    <ModalBaseStat
                        data={ExpectedResults.baseStatsHp}
                    />
                );

        expect(wrapper.contains(<Text testID={'testIdName'}>HP</Text>)).toBe(true);

    })
});