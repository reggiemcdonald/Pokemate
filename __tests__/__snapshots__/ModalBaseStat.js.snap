// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalBaseStat: render Should render a novel base stats data structure without error 1`] = `
<View>
  <View>
    <Text
      testID="testIdName"
    >
      Defense
    </Text>
  </View>
  <RCTScrollView
    data={
      Array [
        Object {
          "change": 1,
          "name": "harden",
        },
        Object {
          "change": 1,
          "name": "withdraw",
        },
        Object {
          "change": 1,
          "name": "defense-curl",
        },
        Object {
          "change": 2,
          "name": "barrier",
        },
        Object {
          "change": 2,
          "name": "acid-armor",
        },
        Object {
          "change": 1,
          "name": "steel-wing",
        },
        Object {
          "change": 1,
          "name": "ancient-power",
        },
        Object {
          "change": 1,
          "name": "silver-wind",
        },
        Object {
          "change": 1,
          "name": "cosmic-power",
        },
        Object {
          "change": 2,
          "name": "iron-defense",
        },
        Object {
          "change": 1,
          "name": "bulk-up",
        },
        Object {
          "change": 1,
          "name": "defend-order",
        },
        Object {
          "change": 1,
          "name": "ominous-wind",
        },
        Object {
          "change": 1,
          "name": "coil",
        },
        Object {
          "change": 3,
          "name": "cotton-guard",
        },
        Object {
          "change": 1,
          "name": "flower-shield",
        },
        Object {
          "change": 1,
          "name": "diamond-storm",
        },
        Object {
          "change": 1,
          "name": "magnetic-flux",
        },
        Object {
          "change": -1,
          "name": "tail-whip",
        },
        Object {
          "change": -1,
          "name": "leer",
        },
        Object {
          "change": -2,
          "name": "screech",
        },
        Object {
          "change": -1,
          "name": "iron-tail",
        },
        Object {
          "change": -1,
          "name": "crunch",
        },
        Object {
          "change": -1,
          "name": "rock-smash",
        },
        Object {
          "change": -1,
          "name": "superpower",
        },
        Object {
          "change": -1,
          "name": "crush-claw",
        },
        Object {
          "change": -1,
          "name": "tickle",
        },
        Object {
          "change": -1,
          "name": "close-combat",
        },
        Object {
          "change": -1,
          "name": "razor-shell",
        },
        Object {
          "change": -1,
          "name": "v-create",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    testId="testIdMoves"
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View>
          <Text />
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View>
          <Text />
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View>
          <Text />
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View>
          <Text />
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View>
          <Text />
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View>
          <Text />
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View>
          <Text />
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View>
          <Text />
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View>
          <Text />
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View>
          <Text />
        </View>
      </View>
      <View
        style={
          Object {
            "height": 0,
          }
        }
      />
    </View>
  </RCTScrollView>
  <RCTScrollView
    data={
      Array [
        Object {
          "name": "bold",
        },
        Object {
          "name": "impish",
        },
        Object {
          "name": "lax",
        },
        Object {
          "name": "relaxed",
        },
        Object {
          "name": "lonely",
        },
        Object {
          "name": "mild",
        },
        Object {
          "name": "gentle",
        },
        Object {
          "name": "hasty",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    testId="testIdNatures"
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View>
          <Text />
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View>
          <Text />
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View>
          <Text />
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View>
          <Text />
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View>
          <Text />
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View>
          <Text />
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View>
          <Text />
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View>
          <Text />
        </View>
      </View>
    </View>
  </RCTScrollView>
  <RCTScrollView
    data={
      Array [
        "https://pokeapi.co/api/v2/characteristic/3/",
        "https://pokeapi.co/api/v2/characteristic/9/",
        "https://pokeapi.co/api/v2/characteristic/15/",
        "https://pokeapi.co/api/v2/characteristic/21/",
        "https://pokeapi.co/api/v2/characteristic/27/",
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    testId="testIdCharacteristics"
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View>
          <Text />
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View>
          <Text />
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View>
          <Text />
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View>
          <Text />
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View>
          <Text />
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`ModalBaseStat: render Should render without error when given a valid base stats data structure 1`] = `
<View>
  <View>
    <Text
      testID="testIdName"
    >
      HP
    </Text>
  </View>
  <RCTScrollView
    data={
      Array [
        Object {
          "change": 1,
          "name": "ancient-power",
        },
        Object {
          "change": 1,
          "name": "silver-wind",
        },
        Object {
          "change": 1,
          "name": "ominous-wind",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    testId="testIdMoves"
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View>
          <Text />
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View>
          <Text />
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View>
          <Text />
        </View>
      </View>
    </View>
  </RCTScrollView>
  <RCTScrollView
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    testId="testIdNatures"
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View />
  </RCTScrollView>
  <RCTScrollView
    data={
      Array [
        "https://pokeapi.co/api/v2/characteristic/1/",
        "https://pokeapi.co/api/v2/characteristic/7/",
        "https://pokeapi.co/api/v2/characteristic/13/",
        "https://pokeapi.co/api/v2/characteristic/19/",
        "https://pokeapi.co/api/v2/characteristic/25/",
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    testId="testIdCharacteristics"
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View>
          <Text />
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View>
          <Text />
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View>
          <Text />
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View>
          <Text />
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View>
          <Text />
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
